Python/Product/Analysis/ProjectEntry.cs

```diff
@@ -84,86 +82,74 @@ IAnalysisCookie cookie
+                    if (_entry._expectedParse == _expected) {
+                        _entry._pendingParse.Set();
                     }
-                    Tree = null;
-                    return;
                 }
+            }
 
-                Tree = newAst;
-                Cookie = newCookie;
-
-                if (_curWaiter != null) {
-                    _curWaiter.Set();
+            public void Complete() {
+                _completed = true;
+                lock (_entry) {
+                    if (_entry._expectedParse == _expected) {
+                        _entry.SetCurrentParse(Tree, Cookie);
+                        _entry._pendingParse.Set();
+                    }
                 }
             }
-
-            OnNewParseTree?.Invoke(this, EventArgs.Empty);
-        }
-
-        internal bool IsVisible(ProjectEntry assigningScope) {
-            return true;
         }
 
-        public void GetTreeAndCookie(out PythonAst tree, out IAnalysisCookie cookie) {
+        public IPythonParse BeginParse() {
+            _pendingParse.Reset();
             lock (this) {
-                tree = Tree;
-                cookie = Cookie;
+                _expectedParse += 1;
+                return new ActivePythonParse(this, _expectedParse);
             }
         }
 
-        public void BeginParsingTree() {
+        public IPythonParse GetCurrentParse() {
             lock (this) {
-                _updatesPending++;
+                return new StaticPythonParse(Tree, Cookie);
             }
         }
 
-        public PythonAst WaitForCurrentTree(int timeout = -1) => WaitForCurrentTree(timeout, out _);
-
-        public PythonAst WaitForCurrentTree(int timeout, out IAnalysisCookie cookie) {
+        public void SetCurrentParse(PythonAst tree, IAnalysisCookie cookie) {
             lock (this) {
-                if (_updatesPending == 0) {
-                    cookie = Cookie;
-                    return Tree;
-                }
-
-                _waiters++;
-                if (_curWaiter == null) {
-                    _curWaiter = Interlocked.Exchange(ref _sharedWaitEvent, null);
-                    if (_curWaiter == null) {
-                        _curWaiter = new ManualResetEventSlim(false);
-                    } else {
-                        _curWaiter.Reset();
-                    }
-                }
+                Tree = tree;
+                Cookie = cookie;
             }
+            OnNewParseTree?.Invoke(this, EventArgs.Empty);
+        }
 
-            bool gotNewTree = _curWaiter.Wait(timeout);
-
-            lock (this) {
-                _waiters--;
-                if (_waiters == 0 &&
-                    Interlocked.CompareExchange(ref _sharedWaitEvent, _curWaiter, null) != null) {
-                    _curWaiter.Dispose();
-                }
-                _curWaiter = null;
+        public IPythonParse WaitForCurrentParse(int timeout = -1) {
+            _pendingParse.Wait(timeout);
```

Comment:
You need to check if `Wait` has returned `true` or `false`

Reply:
Either that or change the definition to say "return the latest after the delay", which is the usual behaviour anyway. But returning `null` here is what the previous version of this did and it helps improve logging.